src/github.com/getlantern/http-proxy-extensions/preprocessor/preprocessor.go

```diff
@@ -0,0 +1,79 @@
+package preprocessor
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"net"
+	"net/http"
+	"sync/atomic"
+
+	"github.com/getlantern/golog"
+	"github.com/getlantern/http-proxy-extensions/mimic"
+)
+
+// StatefulConn is a type of connection that changes it's internal state when its
+// SetState() is called
+type StatefulConn interface {
+	SetState(s http.ConnState)
+}
+
+var log = golog.LoggerFor("preprocessor")
+
+type listener struct {
+	net.Listener
+}
+
+func NewListener(l net.Listener) *listener {
+	listener := &listener{l}
+	return listener
+}
+
+func (sl *listener) Accept() (net.Conn, error) {
+	c, err := sl.Listener.Accept()
+	return &conn{Conn: c, newRequest: 1}, err
+}
+
+type conn struct {
+	net.Conn
+	// ready to handle a new http request when == 1
+	newRequest uint32
+}
+
+func (c *conn) SetState(s http.ConnState) {
+	if s == http.StateIdle {
+		atomic.StoreUint32(&c.newRequest, 1)
+	}
+}
+
+func (c *conn) Read(p []byte) (n int, err error) {
+	if atomic.SwapUint32(&c.newRequest, 0) == 0 {
```

Comment:
We only check the request if it's a new connection, or previous request is completed.